worker/introspection/leases.go

```diff
@@ -0,0 +1,185 @@
+		return
+	}
+	if snapshot.Version != 1 {
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprintf(w, "only understand how to show version 1 snapshots\n")
+		return
+	}
+
+	data := h.translateSnapshot(snapshot)
+
+	q := r.URL.Query()
+	if v := q.Get("model"); v != "" {
+		data = h.filterModel(data, v)
+	}
+	if v := q["app"]; len(v) > 0 {
+		data = h.filterApp(data, v)
+	}
+
+	bytes, err := yaml.Marshal(h.format(data))
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprintf(w, "error: %v\n", err)
+		return
+	}
+	w.Write(bytes)
+}
+
+type leases struct {
+	// TODO: handle pinning
+	controller map[string]leaseInfo
+	models     map[string]map[string]leaseInfo
+}
+
+type leaseInfo struct {
+	holder   string
+	acquired time.Duration
+	expires  time.Duration
+}
+
+func (h leaseHandler) translateSnapshot(snapshot *raftlease.Snapshot) *leases {
+	result := &leases{
+		controller: make(map[string]leaseInfo),
+		models:     make(map[string]map[string]leaseInfo),
+	}
+	now := snapshot.GlobalTime
+	makeLeaseInfo := func(entry raftlease.SnapshotEntry) leaseInfo {
+		acquired := entry.Start.Sub(now).Round(time.Second)
+		return leaseInfo{
+			holder:   entry.Holder,
+			acquired: acquired,
+			// We don't need to round the expiry at this stage because
+			// the entry.Duration is always one minute. Even if this changes
+			// it is unlikely to be a fraction of a second, so we should
+			// still be fine.
+			expires: acquired + entry.Duration,
+		}
+	}
+	for key, value := range snapshot.Entries {
+		switch key.Namespace {
+		case lease.SingularControllerNamespace:
+			result.controller[key.Lease] = makeLeaseInfo(value)
+		case lease.ApplicationLeadershipNamespace:
+			model, ok := result.models[key.ModelUUID]
+			if !ok {
+				model = make(map[string]leaseInfo)
+				result.models[key.ModelUUID] = model
+			}
+			model[key.Lease] = makeLeaseInfo(value)
+		default:
+			logger.Warningf("unknown namespace %q", key.Namespace)
+		}
+	}
+	// TODO: handle pinned.
+	return result
+}
+
+func (h leaseHandler) filterModel(data *leases, partialModelUUID string) *leases {
+	result := &leases{
+		controller: make(map[string]leaseInfo),
+		models:     make(map[string]map[string]leaseInfo),
+	}
+	for modelUUID, value := range data.controller {
+		if strings.Contains(modelUUID, partialModelUUID) {
+			result.controller[modelUUID] = value
+		}
+	}
+	for modelUUID, value := range data.models {
+		if strings.Contains(modelUUID, partialModelUUID) {
+			result.models[modelUUID] = value
+		}
+	}
+
+	return result
+}
+
+func (h leaseHandler) filterApp(data *leases, partialAppNames []string) *leases {
+	result := &leases{
+		models: make(map[string]map[string]leaseInfo),
+	}
+	for modelUUID, apps := range data.models {
+		appInfo := make(map[string]leaseInfo)
+		for appName, value := range apps {
+			// If the appName matches any of the partial names, add it in.
+			for _, partial := range partialAppNames {
+				if strings.Contains(appName, partial) {
```

Comment:
regexp match?
allows for "mysql.*" or ".*mysql" etc
we'd take the partial string and make a regexp of "^" + partial + "$" to make the default match the whole app name
as with partial uuid matches, we want to avoid surprises in the match

Reply:
Adding regexp just adds to problems IMO. Partial string matches are easy to explain, and easy to implement.

You can provide multiple matches, which will be easier than making people think about how to say that in a regexp.